File: 프로그래머스/2/131127. 할인 행사/할인 행사.java
import java.util.HashMap;
import java.util.Map;

public class Solution {
	public int solution(String[] want, int[] number, String[] discount) {
		int answer = 0;
		Map<String, Integer> totalDiscountMap = new HashMap<>();
		Map<String, Integer> wantMap = new HashMap<>();

		for (int i = 0; i < number.length; i++) {
			wantMap.put(want[i], number[i]);
		}

		// 10일간 회원
		int start = 0;
		int end = 9;

		for (int i = 0; i <= end; i++) {
			totalDiscountMap.merge(discount[i], 1, Integer::sum);
		}

		while (end < discount.length) {

			if (totalDiscountMap.equals(wantMap)) {
				answer++;
			}

			if (end == discount.length - 1) {
				break;
			}

			totalDiscountMap.computeIfPresent(discount[start], (k, v) -> (v > 1) ? v - 1 : null);
			start++;
			totalDiscountMap.merge(discount[++end], 1, Integer::sum);
		}

		return answer;
	}
}

File: 프로그래머스/2/131127. 할인 행사/할인 행사.py
from collections import Counter
from typing import List

def solution(want: List[str], number: List[int], discount: List[str]):
    want_map = {}
    
    for wannt, num in zip(want, number):
        want_map[wannt] = num
        
    want_counter = Counter(want_map)
    discount_counter = Counter()
        
    for i in range(9):
        discount_counter.update([discount[i]])
        
    start_idx = 0
    end_idx = 9
    result = 0
    
    while end_idx < len(discount):
        discount_counter.update([discount[end_idx]])
        
        if len(want_counter - discount_counter) == 0: result += 1
        
        discount_counter.subtract([discount[start_idx]])
        start_idx += 1
        end_idx += 1
        
    return result